File: Python3/1042_flower_planting_with_no_adjacent.py
# https://leetcode.cn/problems/flower-planting-with-no-adjacent/

class Solution:
    def gardenNoAdj(self, n, paths):
        """
        Inputs:
            n: int
            paths: list[list[int]]
        Outputs:
            res: list[int]
        """
        graph = defaultdict(set)
        for x, y in paths:
            graph[x].add(y)
            graph[y].add(x)
        
        valid = [[1,1,1,1] for _ in range(n)]
        res = [1 for _ in range(n)]
        for i in range(n):
            for g, check in enumerate(valid[i]):
                if check:
                    res[i] = g + 1
                    for j in graph[i+1]:
                        valid[j-1][g] = 0
                    break
        return res

File: Cpp/1042_flower_planting_with_no_adjacent.cpp
// https://leetcode.cn/problems/flower-planting-with-no-adjacent/

class Solution {
public:
    vector<int> gardenNoAdj(int n, vector<vector<int>>& paths) {
        vector<vector<int>> graph(n);
        for (auto &path: paths) {
            graph[path[0]-1].push_back(path[1]-1);
            graph[path[1]-1].push_back(path[0]-1);
        }
        vector<int> res(n, 1);
        vector<vector<int>> valid(n, vector<int>(4, 1));
        for (int i = 0; i < n; ++i) {
            for (int garden = 0; garden < 4; ++garden) {
                if (valid[i][garden]) {
                    res[i] = garden + 1;
                    for (auto j: graph[i]) {
                        valid[j][garden] = 0;
                    }
                    break;
                }
            }
        }
        return res;
    }
};
